File: 设计模式/程序/2-AbstractFactoryPattern/AbFactoryPattern(Cpp)/src/ShapeFactory.cpp
#include "ShapeFactory.h"
#include "ColorFactory.h"

#include <iostream>
#include <string>


namespace AbstratFactoryPattern
{
    void Rectangle::draw()
    {
        std::cout << "draw a Rectangle" << std::endl;
    }

    void Circle::draw()
    {
        std::cout << "draw a Circle" << std::endl;
    }

    void Square::draw()
    {
        std::cout << "draw a Square" << std::endl;
    }


    /**********************************************************
     * @brief ShapeFactory's GetColor
     * @param shapeType shape need to create
     * @return Shape* object point
     **********************************************************/
    Shape* ShapeFactory::GetShape(const std::string& shapeType) 
    {
        if (shapeType == "Rectangle")
            return new Rectangle();
        else if (shapeType == "Circle")
            return new Circle();
        else if (shapeType == "Square")
            return new Square();
        return nullptr;
    }

    /**********************************************************
     * @brief ShapeFactory's GetColor
     * @param colorType color need to create
     * @return Color* object point
     **********************************************************/
    Color* ShapeFactory::GetColor(const std::string& colorType)
    {
        return nullptr;
    }
}

File: 设计模式/程序/2-AbstractFactoryPattern/AbFactoryPattern(C#)/ShapeFactory.cs
/* 形状工厂 */
using System;


namespace AbstractFactoryPattern
{
    public interface Shape
    {
        void draw();
    }

    public class Circle : Shape
    {
        public void draw()
        {
            Console.WriteLine("Draw a Circle!");
        }
    }

    public class Rectangle : Shape
    {
        public void draw()  
        {
            Console.WriteLine("Draw a Rectangle!"); 
        }
    }

    public class ShapeFactory : AbstractFactory
    {
        public Color GetColor(string colorType)
        { 
            return null;
        }
        public Shape GetShape(string shapeType)
        {
            if (shapeType == "CIRCLE")
                return new Circle();
            else if (shapeType == "RECTANGLE")
                return new Rectangle();
            else 
                return null;
        }
    }
}
